### DISCUSSION TITLE
Transaction indexing with Electrs not working

### DISCUSSION THREAD

**Discussion by evehash** [janoside/btc-rpc-explorer#498] [Q&A] [answered]
Hello there,

I'm running the last version of electrs (0.9.11) and btc-rpc-explorer (v3.3.0) with the following flags enabled:

- BTCEXP_ADDRESS_API=electrum
- BTCEXP_ELECTRUM_SERVERS=tpc://electrs:50001
- BTCEXP_ELECTRUM_TXINDEX=true

Now, if I try to explore a old transaction I can see the in BTC Explorer following:

No results found for query: xxxxx Your node does not have **txindex** enabled. Without it, you can only lookup wallet, mempool, and recently confirmed transactions by their **txid**. Searching for non-wallet transactions that were confirmed more than 3 blocks ago is only possible if the confirmed block height is available.

And this in Electrs:

WARN unknown method blockchain.transaction.get_confirmed_blockhash()

Also, I'm not sure if it's related, but when tbtc-rpc-explorer starts, I can see the following in the terminal:

- btcexp:app txindex check: trying getindexinfo
- btcexp:app txindex check: getindexinfo=()
- btcexp:app txindex check: unavailable

Same problem with electrs (0.9.10)

From my understanding, BTC Explorer has to use electrs as an indexer. That's the idea, right? There is no need to set the txindex in Bitcoin Core.

This is a bug or I'm missing something? Thank you.

**Comment by jsarenik** [comment]
Using `v0.9.11` romanz/electrs@000fc0d with current `btc-rpc-explorer` + few unimportant [patches](https://github.com/janoside/btc-rpc-explorer/commit/1b0dfe60dc) on [be.anyone.eu.org](https://be.anyone.eu.org/) and everything works fine.

Electrs is run like this:

```shell
exec $HOME/bin/electrs --conf $HOME/.config/electrs/main.toml
```

And the main.toml file is this:

```
cookie_file = "/home/nsm/.bitcoin/.cookie"
db_dir = "/home/nsm/.bitcoin/electrs"
network = "bitcoin"
electrum_rpc_addr = "0.0.0.0:60001"
log_filters = "INFO"
index_lookup_limit = 50
ignore_mempool = false
```

Then `btc-rpc-explorer` receives this environment sans user/pass:

```shell
export BTCEXP_PRIVACY_MODE=true
export BTCEXP_HOST=127.0.0.1
export BTCEXP_PORT=3002
export BTCEXP_BITCOIND_USER=user
export BTCEXP_BITCOIND_PASS=bitcoin/share/rpcauth.py-generated
export BTCEXP_ADDRESS_API=electrum
export BTCEXP_ELECTRUM_SERVERS=tcp://127.0.0.1:60001
export BTCEXP_ELECTRUM_TXINDEX=true
export BTCEXP_DEMO=true
export BTCEXP_SLOW_DEVICE_MODE=false
export BTCEXP_NO_RATES=false
export BTCEXP_UI_HOME_PAGE_LATEST_BLOCKS_COUNT=6
export BTCEXP_FILESYSTEM_CACHE_DIR=../../.cache/mainnet
export BTCEXP_UI_TIMEZONE=utc
```

By the way, I always compile `electrs` with `cargo build --no-default-features --locked --release` so that it does not contain Prometheus (monitoring) — as suggested in `electrs/doc/install.md`.

**Reply by evehash** [reply]
Thank you for you answer.

I've a very simillar configuration, but really, I'm a bit confused:

If I go here: https://btcrpcexplorer:3003/admin/dashboard I can see the addressApi is electrum

{
    "privacyMode": true,
    "slowDeviceMode": true,
    "demoSite": false,
    "rpcConcurrency": 3,
    "addressApi": "electrum",
    "ipStackComApiAccessKey": false,
    "mapBoxComApiAccessKey": false,
    "redisCache": false,
    "noInmemoryRpcCache": false
}

Then, if I search for any "old" transaction. I.e: d5a1941a344dab311b5b9175db7cb80d4f99baec6fdbb0bb9f5dbc7293dd9a3c

No results found for query: d5a1941a344dab311b5b9175db7cb80d4f99baec6fdbb0bb9f5dbc7293dd9a3c Your node does not have **txindex** enabled. Without it, you can only lookup wallet, mempool, and recently confirmed transactions by their **txid**. Searching for non-wallet transactions that were confirmed more than 3 blocks ago is only possible if the confirmed block height is available

I don't get it. txindex is a Bitcoin-Core parameter, why is pointing that? why 3 blocks ago and not 4, 5 o 10?

**Comment by evehash** [comment]
@jsarenik Do you have txindex=1 in bitcoind?

Now that I'm reviewing the documentation and the config files... I'm a bit confused.

This is from the config file:

```
BTCEXP_ADDRESS_API=electrum 	# use electrum-protocol servers for address lookups <=== ONLY FOR ADDRESSES?

# Whether to use the Electrum server as an external txindex.
# This is only available in the Electrs implementation. Electrs txindex can serve as a
# full replacement for the Bitcoin Core txindex, but Bitcoin Core's implementation  <==== FULL REPLACEMENT INCLUDING TX?
# is likely faster.
BTCEXP_ELECTRUM_TXINDEX=true
```

And this is from the documentation:
```
In particular, with pruning enabled and/or txindex disabled, the following functionality is altered:

You will only be able to search for mempool, recently confirmed, and wallet transactions by their txid. Searching for non-wallet transactions that were confirmed over 3 blocks ago is only possible if you provide the confirmed block height in addition to the txid.
```

Please @janoside, can you help with this? Can we use Electrs as a replacement and avoid txindex=1 in Bitcoin Core? Right now, if I search a random transaction like: d5a1941a344dab311b5b9175db7cb80d4f99baec6fdbb0bb9f5dbc7293dd9a3c  the app asks for txindex=1.

**Comment by jsarenik** [comment, marked as answer]
@reverse-hash Yes, `txindex=1` on be.anyone.eu.org

**Comment by evehash** [comment]
I was thinking that this parameter has to be 0 since we have elects for indexing. Now it works.

**Reply by jsarenik** [reply]
I fully agree it would be great to work with `txindex=0` and using just electrs. Maybe it works with blockstream/mempool electra. Did not try to run it on my old laptop yet.
